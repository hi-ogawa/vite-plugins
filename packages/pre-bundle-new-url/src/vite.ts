import type { Plugin } from "vite";
import { esbuildPluginNewUrl, esbuildPluginWorkerNewUrl } from "./esbuild";

export function vitePluginPreBundleNewUrl(): Plugin {
  return {
    name: "pre-bundle-new-url",
    config(config, _env) {
      config.cacheDir;
      return {
        optimizeDeps: {
          esbuildOptions: {
            plugins: [
              esbuildPluginWorkerNewUrl(),
              esbuildPluginNewUrl(),
            ],
          },
        },
      };
    },
    configResolved(config) {
      config.cacheDir;
    },
  };
}
