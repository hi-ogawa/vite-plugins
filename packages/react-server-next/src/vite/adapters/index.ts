import type { Plugin } from "vite";

export type AdapterType = "node" | "vercel" | "vercel-edge" | "cloudflare";

export function adapterPlugin(options: {
  adapter?: AdapterType;
  outDir: string;
}): Plugin[] {
  const adapter = options.adapter ?? autoSelectAdapter();
  if (adapter === "node") {
    return [];
  }

  const buildPlugin: Plugin = {
    name: adapterPlugin.name + ":build",
    enforce: "post",
    apply: (_config, env) => !!env.isSsrBuild,
    config() {
      return {
        build: {
          rollupOptions: {
            input: {
              index: `next/vite/adapters/${adapter}/entry`,
            },
          },
        },
      };
    },
    writeBundle: {
      sequential: true,
      async handler() {
        console.log(`▶▶▶ ADAPTER: ${adapter}`);
        if (adapter === "cloudflare") {
          const { build } = await import("./cloudflare/build");
          await build({ outDir: options.outDir });
        }
        if (adapter === "vercel") {
          const { build } = await import("./vercel/build");
          await build({ runtime: "node", outDir: options.outDir });
        }
        if (adapter === "vercel-edge") {
          const { build } = await import("./vercel/build");
          await build({ runtime: "edge", outDir: options.outDir });
        }
      },
    },
  };

  return [buildPlugin];
}

// cf. https://github.com/sveltejs/kit/blob/52e5461b055a104694f276859a7104f58452fab0/packages/adapter-auto/adapters.js
function autoSelectAdapter(): AdapterType {
  if (process.env["VERCEL"]) {
    return "vercel";
  }
  if (process.env["CF_PAGES"]) {
    return "cloudflare";
  }
  return "node";
}
