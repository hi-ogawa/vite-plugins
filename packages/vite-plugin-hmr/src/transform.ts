import type { Node, Program } from "estree";

// traverse export declaration statements based on
// https://github.com/vitejs/vite/blob/fc2bceb09fb65cc6dc843462f51506586251a703/packages/vite/src/node/ssr/ssrTransform.ts#L172

declare module "estree" {
  interface BaseNode {
    start: number;
    end: number;
  }
}

export function analyzeExports(code: string, ast: Program) {
  // extract exported top-level identifiers
  const exportIds: string[] = [];

  // unsupported export usage
  const errors: { node: Node }[] = [];

  for (const node of ast.body) {
    // named exports
    if (node.type === "ExportNamedDeclaration") {
      if (node.declaration) {
        if (
          node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "ClassDeclaration"
        ) {
          //// export function foo() {}
          exportIds.push(node.declaration.id.name);
        } else if (node.declaration.type === "VariableDeclaration") {
          //// export const foo = 1, bar = 2
          if (node.declaration.kind === "const") {
            // rewrite from "const" to "let"
            code = edit(
              code,
              node.declaration.start,
              node.declaration.start + 5,
              "let  "
            );
          }
          for (const decl of node.declaration.declarations) {
            if (decl.id.type === "Identifier") {
              exportIds.push(decl.id.name);
            } else {
              errors.push({ node: decl });
            }
          }
        }
      } else {
        if (node.source) {
          //// export { foo, bar } from './foo'
          errors.push({ node });
        } else {
          //// export { foo, bar }
          // TODO: support by analyzing scope?
          errors.push({ node });
        }
      }
    }

    // default export
    if (node.type === "ExportDefaultDeclaration") {
      if (
        (node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "ClassExpression") &&
        node.declaration.id
      ) {
        //// export default function foo() {}
        //// export default class A {}
        exportIds.push(node.declaration.id.name);
      } else {
        // anonymous default exports
        errors.push({ node });
      }
    }

    //// export * from './foo'
    if (node.type === "ExportAllDeclaration") {
      errors.push({ node });
    }
  }

  return { code, exportIds, errors };
}

function edit(s: string, start: number, end: number, insertee: string) {
  return s.slice(0, start) + insertee + s.slice(end);
}
