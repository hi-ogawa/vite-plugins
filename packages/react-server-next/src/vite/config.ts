// @ts-nocheck
// TODO: this one needs to be `next/vite` when used outside of monorepo
import next from "@hiogawa/react-server-next/vite";
import { defineConfig } from "vite";

export default defineConfig({
  plugins: [next()],
});
