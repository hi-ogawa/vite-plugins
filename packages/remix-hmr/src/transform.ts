import type * as estree from "estree";
import { parseAstAsync } from "vite";

export interface TransformOptions {
  mode: "vite" | "webpack";
  debug: boolean;
}

export type RefreshRuntimeOptions = Pick<TransformOptions, "mode" | "debug">;

export async function transform(code: string, options: TransformOptions) {
  const result = await analyzeCode(code);
  if (result.errors.length || result.entries.length === 0) {
    return;
  }
  const hot = "import.meta.hot";
  const wrap = result.entries
    .map((e) => {
      const key = JSON.stringify(e.hooks.join("/"));
      return `  ${e.id} = $$manager.wrap("${e.id}", ${e.id}, ${key});\n`;
    })
    .join("");
  const footer = `
;import * as $$refresh from "virtual:remix-hmr-runtime";
if (import.meta.hot) {
  (() => import.meta.hot.accept());
  const $$manager = $$refresh.initialize(
    ${hot},
    ${JSON.stringify(options)}
  );

${wrap}
  $$manager.setup();
}
`;
  // no need to manipulate sourcemap since transform only appends
  return result.outCode + footer;
}

//
// extract component declarations
//

// extend types for rollup ast with node position
declare module "estree" {
  interface BaseNode {
    start: number;
    end: number;
  }
}

type ParsedEntry = {
  id: string;
  hooks: string[];
};

const HOOK_CALL_RE = /\b(use\w*)\s*\(/g;
const COMPONENT_RE = /^[A-Z]/;

async function analyzeCode(code: string) {
  const ast = await parseAstAsync(code);
  const errors: unknown[] = [];
  const entries: ParsedEntry[] = [];

  // replace "export const" with "export let"
  let outCode = code;

  // loop exports
  // https://github.com/hi-ogawa/vite-plugins/blob/243064edcf80429be13bee81c0dc1d7bea670f4b/packages/react-server/src/plugin/ast-utils.ts#L14
  for (const node of ast.body) {
    // named exports
    if (node.type === "ExportNamedDeclaration") {
      if (node.declaration) {
        if (node.declaration.type === "FunctionDeclaration") {
          /**
           * export function foo() {}
           */
          if (COMPONENT_RE.test(node.declaration.id.name)) {
            entries.push({
              id: node.declaration.id.name,
              hooks: analyzeFunction(code, node.declaration).hooks,
            });
          }
        } else if (node.declaration.type === "VariableDeclaration") {
          /**
           * export const foo = 1, bar = 2
           */
          if (node.declaration.kind === "const") {
            const start = node.declaration.start;
            outCode = replaceCode(outCode, start, start + 5, "let  ");
          }
          for (const decl of node.declaration.declarations) {
            if (
              decl.id.type === "Identifier" &&
              decl.init &&
              decl.init.type === "ArrowFunctionExpression" &&
              COMPONENT_RE.test(decl.id.name)
            ) {
              entries.push({
                id: decl.id.name,
                hooks: analyzeFunction(code, decl.init).hooks,
              });
            } else {
              errors.push(decl);
            }
          }
        } else {
          errors.push(node);
        }
      } else {
        /**
         * export { foo, bar } from './foo'
         * export { foo, bar as car }
         */
        errors.push(node);
      }
    }

    // default export
    if (node.type === "ExportDefaultDeclaration") {
      /**
       * export default function foo() {}
       */
      if (
        node.declaration.type === "FunctionDeclaration" &&
        node.declaration.id
      ) {
        entries.push({
          id: node.declaration.id.name,
          hooks: analyzeFunction(code, node.declaration).hooks,
        });
      } else {
        errors.push(node);
      }
    }

    /**
     * export * from './foo'
     */
    if (node.type === "ExportAllDeclaration") {
      errors.push(node);
    }

    /**
     * function foo() {}
     */
    if (
      node.type === "FunctionDeclaration" &&
      COMPONENT_RE.test(node.id.name)
    ) {
      entries.push({
        id: node.id.name,
        hooks: analyzeFunction(code, node).hooks,
      });
    }

    /**
     * const foo = 1, bar = 2
     */
    if (node.type === "VariableDeclaration") {
      if (node.kind === "const") {
        const start = node.start;
        outCode = replaceCode(outCode, start, start + 5, "let  ");
      }
      for (const decl of node.declarations) {
        // TODO: FunctionExpression
        if (
          decl.id.type === "Identifier" &&
          decl.init &&
          decl.init.type === "ArrowFunctionExpression" &&
          COMPONENT_RE.test(decl.id.name)
        ) {
          entries.push({
            id: decl.id.name,
            hooks: analyzeFunction(code, decl.init).hooks,
          });
        }
      }
    }
  }

  return {
    entries,
    outCode,
    errors,
  };
}

function analyzeFunction(
  code: string,
  node:
    | estree.FunctionDeclaration
    | estree.ArrowFunctionExpression
    | estree.MaybeNamedFunctionDeclaration,
) {
  // we could do this runtime via `fn.toString()`,
  // but that will impact performance on each render.
  const bodyCode = code.slice(node.body.start, node.body.end);
  const matches = bodyCode.matchAll(HOOK_CALL_RE);
  const hooks = [...matches].map((m) => m[1]!);
  return { hooks };
}

function replaceCode(
  code: string,
  start: number,
  end: number,
  content: string,
) {
  return code.slice(0, start) + content + code.slice(end);
}
