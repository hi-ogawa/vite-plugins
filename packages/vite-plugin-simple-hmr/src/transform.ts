import type { Node, Program } from "estree";
import MagicString from "magic-string";
import { parseAstAsync } from "vite";
import { name as packageName } from "../package.json";

export async function hmrTransform(code: string) {
  const magic = new MagicString(code);

  const ast = await parseAstAsync(code);
  const result = analyzeExports(magic, ast as any);

  if (result.errors.length > 0) {
    const node = result.errors[0]!.node;
    const message = "unsupported usage: " + code.slice(node.start, node.end);
    magic.append("\n" + generateFooterUnsupported(message));
  } else {
    magic.append("\n" + generateFooter(result.exportIds));
  }

  return {
    code: magic.toString(),
    map: magic.generateMap(),
  };
}

function generateFooter(names: string[]) {
  const parts = names.map(
    (name) => `
  $$registry.exports["${name}"] = {
    value: ${name},
    update: ($$next) => {
      ${name} = $$next;
    }
  };
`,
  );

  // requires dummy "hot.accept" for vite to detect
  return `
if (import.meta.env.SSR && import.meta.hot) {
  const $$hmr = await import("${packageName}/runtime");
  const $$registry = $$hmr.createRegistry();

${parts.join("\n")}

  $$hmr.setupHot(import.meta.hot, $$registry);
  import.meta.hot.accept;
}
`;
}

// always invalidate on unsupported usage
function generateFooterUnsupported(message: string) {
  return `
if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate(${JSON.stringify(message)})
  });
}
`;
}

// traverse export declaration statements based on
// https://github.com/vitejs/vite/blob/fc2bceb09fb65cc6dc843462f51506586251a703/packages/vite/src/node/ssr/ssrTransform.ts#L172

declare module "estree" {
  interface BaseNode {
    start: number;
    end: number;
  }
}

export function analyzeExports(code: MagicString, ast: Program) {
  // extract exported top-level identifiers
  const exportIds: string[] = [];

  // unsupported export usage
  const errors: { node: Node }[] = [];

  for (const node of ast.body) {
    // named exports
    if (node.type === "ExportNamedDeclaration") {
      if (node.declaration) {
        if (
          node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "ClassDeclaration"
        ) {
          /**
           * export function foo() {}
           */
          exportIds.push(node.declaration.id.name);
        } else if (node.declaration.type === "VariableDeclaration") {
          /**
           * export const foo = 1, bar = 2
           */
          if (node.declaration.kind === "const") {
            // rewrite from "const" to "let"
            code.remove(node.declaration.start, node.declaration.start + 5);
            code.appendLeft(node.declaration.start, "let");
          }
          for (const decl of node.declaration.declarations) {
            if (decl.id.type === "Identifier") {
              exportIds.push(decl.id.name);
            } else {
              errors.push({ node: decl });
            }
          }
        }
      } else {
        if (node.source) {
          /**
           * export { foo, bar } from './foo'
           */
        } else {
          /**
           * export { foo, bar }
           */
          // TODO: support by analyzing scope? or just rewrite all top level `const` into `let`?
          errors.push({ node });
        }
      }
    }

    // default export
    if (node.type === "ExportDefaultDeclaration") {
      if (
        (node.declaration.type === "FunctionDeclaration" ||
          node.declaration.type === "ClassExpression") &&
        node.declaration.id
      ) {
        /**
         * export default function foo() {}
         * export default class A {}
         */
        exportIds.push(node.declaration.id.name);
      } else {
        // anonymous default exports
        errors.push({ node });
      }
    }

    /**
     * export * from './foo'
     */
    if (node.type === "ExportAllDeclaration") {
    }
  }

  return { exportIds, errors };
}
