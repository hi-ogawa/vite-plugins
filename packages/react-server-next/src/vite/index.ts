import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import {
  type ReactServerPluginOptions,
  vitePluginReactServer,
} from "@hiogawa/react-server/plugin";
import {
  vitePluginLogger,
  vitePluginSsrMiddleware,
} from "@hiogawa/vite-plugin-ssr-middleware";
import react from "@vitejs/plugin-react-swc";
import { type Plugin, type PluginOption, transformWithEsbuild } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import { type AdapterType, adapterPlugin } from "./adapters";

export type ReactServerNextPluginOptions = {
  adapter?: AdapterType;
};

export default function vitePluginReactServerNext(
  options?: ReactServerPluginOptions & ReactServerNextPluginOptions,
): PluginOption {
  const outDir = options?.outDir ?? "dist";
  return [
    react(),
    nextJsxPlugin(),
    tsconfigPaths(),
    vitePluginReactServer({
      ...options,
      routeDir: options?.routeDir ?? "app",
      plugins: [nextJsxPlugin(), tsconfigPaths(), ...(options?.plugins ?? [])],
    }),
    vitePluginLogger(),
    vitePluginSsrMiddleware({
      entry: "next/vite/entry-ssr",
      preview: path.resolve(outDir, "server", "index.js"),
    }),
    adapterPlugin({ adapter: options?.adapter, outDir }),
    appFaviconPlugin(),
    {
      name: "next-exclude-optimize",
      config: () => ({
        ssr: {
          noExternal: ["next"],
        },
        optimizeDeps: {
          exclude: ["next"],
        },
      }),
    },
  ];
}

function nextJsxPlugin(): Plugin {
  return {
    name: nextJsxPlugin.name,
    // need to override next.js's default tsconfig `jsx: preserve`
    config: () => ({
      esbuild: { jsx: "automatic" },
      optimizeDeps: {
        esbuildOptions: { jsx: "automatic", loader: { ".js": "jsx" } },
      },
    }),
    // manually use esbuild transform to support jsx in js
    // TODO: try using vite-plugin-react-swc and parserConfig to support HMR
    // https://github.com/vitejs/vite-plugin-react-swc?tab=readme-ov-file#parserconfig
    async transform(code, id, _options) {
      if (!id.includes("/node_modules/") && id.endsWith(".js")) {
        return transformWithEsbuild(code, id, {
          loader: "jsx",
          jsx: "automatic",
        });
      }
      return;
    },
  };
}

/** @todo https://nextjs.org/docs/app/api-reference/file-conventions/metadata/app-icons */
function appFaviconPlugin(): Plugin {
  // not sure what exactly Next.js does.
  // for now, let's do quick workaround for app/favicon.ico
  return {
    name: appFaviconPlugin.name,
    apply: (_config, env) => env.command === "serve" || !env.isSsrBuild,
    configureServer(server) {
      if (existsSync("app/favicon.ico")) {
        server.middlewares.use((req, _res, next) => {
          const url = new URL(req.url || "", "https://tmp.local");
          if (url.pathname === "/favicon.ico") {
            req.url = "/app/favicon.ico";
          }
          next();
        });
      }
    },
    generateBundle() {
      if (existsSync("app/favicon.ico")) {
        this.emitFile({
          type: "asset",
          fileName: "favicon.ico",
          source: readFileSync("app/favicon.ico"),
        });
      }
    },
  };
}
