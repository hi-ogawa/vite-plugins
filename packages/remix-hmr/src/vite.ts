import { type FilterPattern, type Plugin, createFilter } from "vite";
import { transform } from "./transform";

export default function vitePluginRemixHmr(options?: {
  include?: FilterPattern;
  exclude?: FilterPattern;
}): Plugin {
  const filter = createFilter(
    options?.include ?? /\.[tj]sx$/,
    options?.exclude ?? /\/node_modules\//,
  );
  return {
    name: "remix-hmr",
    apply: "serve",
    applyToEnvironment: (environment) => environment.name === "client",
    transform: {
      handler(code, id) {
        if (filter(id)) {
          return transform(code, {
            mode: "vite",
            debug: true,
          });
        }
      },
    },
    resolveId: {
      handler(source) {
        if (source === "virtual:remix-hmr-runtime") {
          return this.resolve("./runtime.js", import.meta.filename);
        }
      },
    },
  };
}
