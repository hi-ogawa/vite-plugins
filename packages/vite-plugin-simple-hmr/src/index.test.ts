import MagicString from "magic-string";
import { parseAstAsync } from "vite";
import { describe, expect, it } from "vitest";
import { analyzeExports, hmrTransform } from "./transform";

describe(hmrTransform, () => {
  const examples = [
    [
      "basic",
      `
export function Test1() {}

export let Test2 = () => {}

export /*some*/ const /*comment*/ Test3 = () => {}

export default function Test4() {}
`,
    ],
    [
      "separate named export (unsupported)",
      `
const SomeNamed = () => {};
export { SomeNamed };
`,
    ],
    [
      "separate default export (unsupported)",
      `
const SomeDeafult = () => {};
export default SomeDefault;
`,
    ],
    [
      "anonymous default export (unsupported)",
      `
export default () => {};
`,
    ],
  ] as const;

  examples.forEach(([title, code], i) => {
    describe(`${i} - ${title}`, () => {
      it(analyzeExports, async () => {
        const ast = await parseAstAsync(code);
        const { exportIds, errors } = analyzeExports(
          new MagicString(code),
          ast as any,
        );
        expect({
          exportIds,
          errors: errors.map((e) => code.slice(e.node.start, e.node.end)),
        }).toMatchSnapshot();
      });

      it(hmrTransform, async () => {
        expect(await hmrTransform(code)).toMatchSnapshot();
      });
    });
  });
});
