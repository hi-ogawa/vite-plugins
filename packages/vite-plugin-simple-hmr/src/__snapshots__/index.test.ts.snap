// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hmrTransform > 0 - basic > analyzeExports 1`] = `
{
  "errors": [],
  "exportIds": [
    "Test1",
    "Test2",
    "Test3",
    "Test4",
  ],
}
`;

exports[`hmrTransform > 0 - basic > hmrTransform 1`] = `
{
  "code": "
export function Test1() {}

export let Test2 = () => {}

export /*some*/ let /*comment*/ Test3 = () => {}

export default function Test4() {}


if (import.meta.env.SSR && import.meta.hot) {
  const $$hmr = await import("@hiogawa/vite-plugin-simple-hmr/runtime");
  const $$registry = $$hmr.createRegistry();


  $$registry.exports["Test1"] = {
    value: Test1,
    update: ($$next) => {
      Test1 = $$next;
    }
  };


  $$registry.exports["Test2"] = {
    value: Test2,
    update: ($$next) => {
      Test2 = $$next;
    }
  };


  $$registry.exports["Test3"] = {
    value: Test3,
    update: ($$next) => {
      Test3 = $$next;
    }
  };


  $$registry.exports["Test4"] = {
    value: Test4,
    update: ($$next) => {
      Test4 = $$next;
    }
  };


  $$hmr.setupHot(import.meta.hot, $$registry);
  import.meta.hot.accept;
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA,mBAAqB;AACrB;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`hmrTransform > 1 - separate named export (unsupported) > analyzeExports 1`] = `
{
  "errors": [
    "export { SomeNamed };",
  ],
  "exportIds": [],
}
`;

exports[`hmrTransform > 1 - separate named export (unsupported) > hmrTransform 1`] = `
{
  "code": "
const SomeNamed = () => {};
export { SomeNamed };


if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate("unsupported usage: export { SomeNamed };")
  });
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`hmrTransform > 2 - separate default export (unsupported) > analyzeExports 1`] = `
{
  "errors": [
    "export default SomeDefault;",
  ],
  "exportIds": [],
}
`;

exports[`hmrTransform > 2 - separate default export (unsupported) > hmrTransform 1`] = `
{
  "code": "
const SomeDeafult = () => {};
export default SomeDefault;


if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate("unsupported usage: export default SomeDefault;")
  });
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`hmrTransform > 3 - anonymous default export (unsupported) > analyzeExports 1`] = `
{
  "errors": [
    "export default () => {};",
  ],
  "exportIds": [],
}
`;

exports[`hmrTransform > 3 - anonymous default export (unsupported) > hmrTransform 1`] = `
{
  "code": "
export default () => {};


if (import.meta.env.SSR && import.meta.hot) {
  import.meta.hot.accept(() => {
    import.meta.hot.invalidate("unsupported usage: export default () => {};")
  });
}
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
